
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://Yichen-Si.github.io/ficture/localrun_legacy/">
      
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.17">
    
    
      
        <title>Real data process in a local linux machine - ficture</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.bcfcd587.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#real-data-process-in-a-local-linux-machine" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="ficture" class="md-header__button md-logo" aria-label="ficture" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ficture
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Real data process in a local linux machine
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8v2m9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1 0 1.71-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/seqscope/ficture" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/ficture
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../install/" class="md-tabs__link">
        
  
    
  
  Install

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../quickstart/" class="md-tabs__link">
        
  
    
  
  Quickstart

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../localrun/" class="md-tabs__link">
        
  
    
  
  Small run

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../run/" class="md-tabs__link">
        
  
    
  
  Run

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../format_input/" class="md-tabs__link">
          
  
    
  
  Format input

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="ficture" class="md-nav__button md-logo" aria-label="ficture" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    ficture
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/seqscope/ficture" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/ficture
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Install
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quickstart
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../localrun/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Small run
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../run/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Format input
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Format input
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../format_input/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../format_input/cosmx/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CosMx SMI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../format_input/visiumHD/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visium HD
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../format_input/xenium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Xenium
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../format_input/vizgen/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Vizgen MERSCOPE
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    <span class="md-ellipsis">
      Input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#prepare-environment" class="md-nav__link">
    <span class="md-ellipsis">
      Prepare environment
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#process" class="md-nav__link">
    <span class="md-ellipsis">
      Process
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Process">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      Preprocessing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#model-training" class="md-nav__link">
    <span class="md-ellipsis">
      Model training
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pixel-level-decoding" class="md-nav__link">
    <span class="md-ellipsis">
      Pixel level decoding
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#optional-post-processing" class="md-nav__link">
    <span class="md-ellipsis">
      Optional post-processing
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="real-data-process-in-a-local-linux-machine">Real data process in a local linux machine<a class="headerlink" href="#real-data-process-in-a-local-linux-machine" title="Permanent link">&para;</a></h1>
<p>We take a small sub-region of Vizgen MERSCOPE mouse liver data as an example.</p>
<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h2>
<h3 id="input">Input<a class="headerlink" href="#input" title="Permanent link">&para;</a></h3>
<p><div class="highlight"><pre><span></span><code>examples/data/transcripts.tsv.gz
examples/data/feature.clean.tsv.gz
</code></pre></div>
(All filese are tab-delimited text files unless specified otherwise.)</p>
<p>See the following explanation for each file. If you have trouble, try <a href="../format_input/">Format input</a> to see some examples of formating raw output from different platforms.</p>
<p><strong>Transcripts</strong></p>
<p>One file contains the molecular or pixel level information, the required columns are <code>X</code>, <code>Y</code>, <code>gene</code>, and <code>Count</code>. (There could be other columns in the file which would be ignored.)</p>
<p>The coordinates <code>(X, Y)</code> can be float or integer numbers in arbitrary units, but if it is not in the unit of <span class="arithmatex">\(\mu m\)</span> we would need to specify the translation ratio later.</p>
<p>The file has to be <strong>sorted</strong> by one of the coordinates. (Usually it is the longer axis, but it does not matter if the tissue area is not super asymmetric.)</p>
<p><code>Count</code> (could be any other name) is the number of transcripts for the specified <code>gene</code> observed at the coordinate. For imaging based technologies where each molecule has its unique coordinates, <code>Count</code> could be always 1.</p>
<p><strong>Gene list</strong></p>
<p>Another file contains the (unique) names of genes that should be used in analysis. The required columns is just <code>gene</code> (including the header), the naming of genes should match the <code>gene</code> column in the transcript file. If your data contain negative control probes or if you would like to remove certain genes this is where you can specify. (If you would like to use all genes present in your input transcript file the gene list is not necessary, but you would need to modify the command in <code>generic_III.sh</code> to remove the argument <code>--feature</code> )</p>
<p><strong>Meta data</strong></p>
<p>We also prefer to keep a file listing the min and max of the coordinates (this is primarily for visuaizing very big tissue region where we do not read all data at once but would want to know the image dimension). The unit of the coordinates is micrometer.
<div class="highlight"><pre><span></span><code>examples/data/coordinate_minmax.tsv
</code></pre></div></p>
<h3 id="prepare-environment">Prepare environment<a class="headerlink" href="#prepare-environment" title="Permanent link">&para;</a></h3>
<p>Suppose you have installed dependencies following <a href="../install/">Install</a>. This document assume you are on the <code>stable</code> branch.</p>
<p>Activate your virtual environmnet if needed
<div class="highlight"><pre><span></span><code>source venv/with/requirements/installed/bin/activate
</code></pre></div></p>
<h2 id="process">Process<a class="headerlink" href="#process" title="Permanent link">&para;</a></h2>
<p>Specify the base directory that contains the input data
<div class="highlight"><pre><span></span><code><span class="nv">path</span><span class="o">=</span>examples/data
</code></pre></div></p>
<p>Data specific setup:</p>
<p><code>mu_scale</code> is the ratio between <span class="arithmatex">\(\mu m\)</span> and the unit used in the transcript coordinates. For example, if the coordinates are sotred in <code>nm</code> this number should be <code>1000</code>.</p>
<p><code>key</code> is the column name in the transcripts file corresponding to the gene counts (<code>Count</code> in our example). <code>major_axis</code> specify which axis the transcript file is sorted by.</p>
<div class="highlight"><pre><span></span><code><span class="nv">mu_scale</span><span class="o">=</span><span class="m">1</span><span class="w"> </span><span class="c1"># If your data&#39;s coordinates are already in micrometer</span>
<span class="nv">key</span><span class="o">=</span>Count
<span class="nv">major_axis</span><span class="o">=</span>Y<span class="w"> </span><span class="c1"># If your data is sorted by the Y-axis</span>
<span class="nv">gitpath</span><span class="o">=</span>path/to/ficture<span class="w"> </span><span class="c1"># path to this repository</span>
</code></pre></div>
<h3 id="preprocessing">Preprocessing<a class="headerlink" href="#preprocessing" title="Permanent link">&para;</a></h3>
<p>Create pixel minibatches (<code>${path}/batched.matrix.tsv.gz</code>)
<div class="highlight"><pre><span></span><code><span class="nv">batch_size</span><span class="o">=</span><span class="m">500</span>
<span class="nv">batch_buff</span><span class="o">=</span><span class="m">30</span>
<span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/transcripts.tsv.gz
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/batched.matrix.tsv.gz
<span class="nv">batch</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/batched.matrix.tsv

python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/make_spatial_minibatch.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">batch</span><span class="si">}</span><span class="w"> </span>--mu_scale<span class="w"> </span><span class="si">${</span><span class="nv">mu_scale</span><span class="si">}</span><span class="w"> </span>--batch_size<span class="w"> </span><span class="si">${</span><span class="nv">batch_size</span><span class="si">}</span><span class="w"> </span>--batch_buff<span class="w"> </span><span class="si">${</span><span class="nv">batch_buff</span><span class="si">}</span><span class="w"> </span>--major_axis<span class="w"> </span><span class="si">${</span><span class="nv">major_axis</span><span class="si">}</span>

sort<span class="w"> </span>-S<span class="w"> </span>4G<span class="w"> </span>-k2,2n<span class="w"> </span>-k1,1g<span class="w"> </span><span class="si">${</span><span class="nv">batch</span><span class="si">}</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>gzip<span class="w"> </span>-c<span class="w"> </span>&gt;<span class="w"> </span><span class="si">${</span><span class="nv">batch</span><span class="si">}</span>.gz
rm<span class="w"> </span><span class="si">${</span><span class="nv">batch</span><span class="si">}</span>
</code></pre></div></p>
<p>Prepare training minibatches, only need to run once if you plan to fit multiple models (say with different number of factors)
<div class="highlight"><pre><span></span><code><span class="c1"># Prepare training minibatches, only need to run once if you plan to fit multiple models (say with different number of factors)</span>
<span class="nv">train_width</span><span class="o">=</span><span class="m">12</span><span class="w"> </span><span class="c1"># \sqrt{3} x the side length of the hexagon (um)</span>
<span class="nv">min_ct_per_unit</span><span class="o">=</span><span class="m">50</span>
<span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/transcripts.tsv.gz
<span class="nv">out</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/hexagon.d_<span class="si">${</span><span class="nv">train_width</span><span class="si">}</span>.tsv
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/make_dge_univ.py<span class="w"> </span>--key<span class="w"> </span><span class="si">${</span><span class="nv">key</span><span class="si">}</span><span class="w"> </span>--count_header<span class="w"> </span><span class="si">${</span><span class="nv">key</span><span class="si">}</span><span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">out</span><span class="si">}</span><span class="w"> </span>--hex_width<span class="w"> </span><span class="si">${</span><span class="nv">train_width</span><span class="si">}</span><span class="w"> </span>--n_move<span class="w"> </span><span class="m">2</span><span class="w"> </span>--min_ct_per_unit<span class="w"> </span><span class="si">${</span><span class="nv">min_ct_per_unit</span><span class="si">}</span><span class="w"> </span>--mu_scale<span class="w"> </span><span class="si">${</span><span class="nv">mu_scale</span><span class="si">}</span><span class="w"> </span>--precision<span class="w"> </span><span class="m">2</span><span class="w"> </span>--major_axis<span class="w"> </span><span class="si">${</span><span class="nv">major_axis</span><span class="si">}</span>

sort<span class="w"> </span>-S<span class="w"> </span>4G<span class="w"> </span>-k1,1n<span class="w"> </span><span class="si">${</span><span class="nv">out</span><span class="si">}</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>gzip<span class="w"> </span>-c<span class="w"> </span>&gt;<span class="w"> </span><span class="si">${</span><span class="nv">out</span><span class="si">}</span>.gz<span class="w"> </span><span class="c1"># Shuffle hexagons</span>
rm<span class="w"> </span><span class="si">${</span><span class="nv">out</span><span class="si">}</span>
</code></pre></div></p>
<h3 id="model-training">Model training<a class="headerlink" href="#model-training" title="Permanent link">&para;</a></h3>
<p>Parameters for initializing the model
<div class="highlight"><pre><span></span><code><span class="nv">nFactor</span><span class="o">=</span><span class="m">12</span><span class="w"> </span><span class="c1"># Number of factors</span>
<span class="nv">sliding_step</span><span class="o">=</span><span class="m">2</span>
<span class="nv">train_nEpoch</span><span class="o">=</span><span class="m">3</span>
<span class="c1"># train_width=12 # should be the same as used in the above step</span>
<span class="nv">model_id</span><span class="o">=</span>nF<span class="si">${</span><span class="nv">nFactor</span><span class="si">}</span>.d_<span class="si">${</span><span class="nv">train_width</span><span class="si">}</span><span class="w"> </span><span class="c1"># An identifier kept in output file names</span>
<span class="nv">min_ct_per_feature</span><span class="o">=</span><span class="m">20</span><span class="w"> </span><span class="c1"># Ignore genes with total count \&lt; 20</span>
<span class="nv">R</span><span class="o">=</span><span class="m">10</span><span class="w"> </span><span class="c1"># We use R random initializations and pick one to fit the full model</span>
<span class="nv">thread</span><span class="o">=</span><span class="m">4</span><span class="w"> </span><span class="c1"># Number of threads to use</span>
</code></pre></div></p>
<p>Initialize the model
<div class="highlight"><pre><span></span><code><span class="c1"># parameters</span>
<span class="nv">min_ct_per_unit_fit</span><span class="o">=</span><span class="m">20</span>
<span class="nv">cmap_name</span><span class="o">=</span><span class="s2">&quot;turbo&quot;</span>
<span class="nv">fit_nmove</span><span class="o">=</span><span class="k">$((</span><span class="nv">fit_width</span><span class="o">/</span><span class="nv">anchor_res</span><span class="k">))</span>

<span class="c1"># output identifiers</span>
<span class="nv">model_id</span><span class="o">=</span>nF<span class="si">${</span><span class="nv">nFactor</span><span class="si">}</span>.d_<span class="si">${</span><span class="nv">train_width</span><span class="si">}</span>
<span class="nv">output_id</span><span class="o">=</span><span class="si">${</span><span class="nv">model_id</span><span class="si">}</span>
<span class="nv">anchor_info</span><span class="o">=</span>prj_<span class="si">${</span><span class="nv">fit_width</span><span class="si">}</span>.r_<span class="si">${</span><span class="nv">anchor_res</span><span class="si">}</span>
<span class="nv">output_path</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/analysis/<span class="si">${</span><span class="nv">model_id</span><span class="si">}</span>
<span class="nv">figure_path</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/figure
<span class="k">if</span><span class="w"> </span><span class="o">[</span><span class="w"> </span>!<span class="w"> </span>-d<span class="w"> </span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span><span class="s2">/sub&quot;</span><span class="w"> </span><span class="o">]</span><span class="p">;</span><span class="w"> </span><span class="k">then</span>
<span class="w">    </span>mkdir<span class="w"> </span>-p<span class="w"> </span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/sub
<span class="k">fi</span>


<span class="c1"># input files</span>
<span class="nv">hexagon</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/hexagon.d_<span class="si">${</span><span class="nv">train_width</span><span class="si">}</span>.tsv.gz
<span class="nv">pixel</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/transcripts.tsv.gz
<span class="nv">feature</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/feature.clean.tsv.gz
<span class="c1"># output</span>
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>
<span class="nv">model</span><span class="o">=</span><span class="si">${</span><span class="nv">output</span><span class="si">}</span>.model.p

<span class="c1"># Fit model</span>
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/init_model_selection.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">hexagon</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--feature<span class="w"> </span><span class="si">${</span><span class="nv">feature</span><span class="si">}</span><span class="w"> </span>--nFactor<span class="w"> </span><span class="si">${</span><span class="nv">nFactor</span><span class="si">}</span><span class="w"> </span>--epoch<span class="w"> </span><span class="si">${</span><span class="nv">train_nEpoch</span><span class="si">}</span><span class="w"> </span>--epoch_id_length<span class="w"> </span><span class="m">2</span><span class="w"> </span>--unit_attr<span class="w"> </span>X<span class="w"> </span>Y<span class="w"> </span>--key<span class="w"> </span><span class="si">${</span><span class="nv">key</span><span class="si">}</span><span class="w"> </span>--min_ct_per_feature<span class="w"> </span><span class="si">${</span><span class="nv">min_ct_per_feature</span><span class="si">}</span><span class="w"> </span>--test_split<span class="w"> </span><span class="m">0</span>.5<span class="w"> </span>--R<span class="w"> </span><span class="si">${</span><span class="nv">R</span><span class="si">}</span><span class="w"> </span>--thread<span class="w"> </span><span class="si">${</span><span class="nv">thread</span><span class="si">}</span>

<span class="c1"># Choose color</span>
<span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.fit_result.tsv.gz
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>
<span class="nv">cmap</span><span class="o">=</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.rgb.tsv
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/choose_color.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--cmap_name<span class="w"> </span><span class="si">${</span><span class="nv">cmap_name</span><span class="si">}</span>

<span class="c1"># Coarse plot for inspection</span>
<span class="nv">cmap</span><span class="o">=</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.rgb.tsv
<span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.fit_result.tsv.gz
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.coarse
<span class="nv">fillr</span><span class="o">=</span><span class="k">$((</span><span class="nv">fit_width</span><span class="o">/</span><span class="m">2</span><span class="o">+</span><span class="m">1</span><span class="k">))</span>
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/plot_base.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--fill_range<span class="w"> </span><span class="si">${</span><span class="nv">fillr</span><span class="si">}</span><span class="w"> </span>--color_table<span class="w"> </span><span class="si">${</span><span class="nv">cmap</span><span class="si">}</span><span class="w"> </span>--plot_um_per_pixel<span class="w"> </span><span class="m">1</span><span class="w"> </span>--plot_discretized
</code></pre></div></p>
<h3 id="pixel-level-decoding">Pixel level decoding<a class="headerlink" href="#pixel-level-decoding" title="Permanent link">&para;</a></h3>
<p>Parameters for pixel level decoding
<div class="highlight"><pre><span></span><code><span class="nv">fit_width</span><span class="o">=</span><span class="m">12</span><span class="w"> </span><span class="c1"># Often equal or smaller than train_width (um)</span>
<span class="nv">anchor_res</span><span class="o">=</span><span class="m">4</span><span class="w"> </span><span class="c1"># Distance between adjacent anchor points (um)</span>
<span class="nv">radius</span><span class="o">=</span><span class="k">$((</span><span class="nv">$anchor_res</span><span class="o">+</span><span class="m">1</span><span class="k">))</span>
<span class="nv">anchor_info</span><span class="o">=</span>prj_<span class="si">${</span><span class="nv">fit_width</span><span class="si">}</span>.r_<span class="si">${</span><span class="nv">anchor_res</span><span class="si">}</span><span class="w"> </span><span class="c1"># An identifier</span>
<span class="nv">coor</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/coordinate_minmax.tsv
<span class="nv">cmap</span><span class="o">=</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.rgb.tsv
</code></pre></div></p>
<div class="highlight"><pre><span></span><code><span class="c1"># Transform</span>
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.<span class="si">${</span><span class="nv">anchor_info</span><span class="si">}</span>
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/transform_univ.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">pixel</span><span class="si">}</span><span class="w"> </span>--output_pref<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--model<span class="w"> </span><span class="si">${</span><span class="nv">model</span><span class="si">}</span><span class="w"> </span>--key<span class="w"> </span><span class="si">${</span><span class="nv">key</span><span class="si">}</span><span class="w"> </span>--major_axis<span class="w"> </span><span class="si">${</span><span class="nv">major_axis</span><span class="si">}</span><span class="w"> </span>--hex_width<span class="w"> </span><span class="si">${</span><span class="nv">fit_width</span><span class="si">}</span><span class="w"> </span>--n_move<span class="w"> </span><span class="si">${</span><span class="nv">fit_nmove</span><span class="si">}</span><span class="w"> </span>--min_ct_per_unit<span class="w"> </span><span class="si">${</span><span class="nv">min_ct_per_unit_fit</span><span class="si">}</span><span class="w"> </span>--mu_scale<span class="w"> </span><span class="si">${</span><span class="nv">mu_scale</span><span class="si">}</span><span class="w"> </span>--thread<span class="w"> </span><span class="si">${</span><span class="nv">thread</span><span class="si">}</span><span class="w"> </span>--precision<span class="w"> </span><span class="m">2</span>

<span class="c1"># Pixel level decoding &amp; visualization</span>
<span class="nv">prefix</span><span class="o">=</span><span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.decode.<span class="si">${</span><span class="nv">anchor_info</span><span class="si">}</span>_<span class="si">${</span><span class="nv">radius</span><span class="si">}</span>
<span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">path</span><span class="si">}</span>/batched.matrix.tsv.gz
<span class="nv">anchor</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.<span class="si">${</span><span class="nv">anchor_info</span><span class="si">}</span>.fit_result.tsv.gz
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>
<span class="c1"># Output only a few top factors per pixel</span>
<span class="nv">topk</span><span class="o">=</span><span class="m">3</span><span class="w"> </span><span class="c1"># Fix for now</span>
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/slda_decode.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--model<span class="w"> </span><span class="si">${</span><span class="nv">model</span><span class="si">}</span><span class="w"> </span>--anchor<span class="w"> </span><span class="si">${</span><span class="nv">anchor</span><span class="si">}</span><span class="w"> </span>--anchor_in_um<span class="w"> </span>--neighbor_radius<span class="w"> </span><span class="si">${</span><span class="nv">radius</span><span class="si">}</span><span class="w"> </span>--mu_scale<span class="w"> </span><span class="si">${</span><span class="nv">mu_scale</span><span class="si">}</span><span class="w"> </span>--key<span class="w"> </span><span class="si">${</span><span class="nv">key</span><span class="si">}</span><span class="w"> </span>--precision<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span>--lite_topk_output_pixel<span class="w"> </span><span class="si">${</span><span class="nv">topk</span><span class="si">}</span><span class="w"> </span>--lite_topk_output_anchor<span class="w"> </span><span class="si">${</span><span class="nv">topk</span><span class="si">}</span><span class="w"> </span>--thread<span class="w"> </span><span class="si">${</span><span class="nv">thread</span><span class="si">}</span>
</code></pre></div>
<h3 id="optional-post-processing">Optional post-processing<a class="headerlink" href="#optional-post-processing" title="Permanent link">&para;</a></h3>
<p>The following is not strictly necessary but it generates summary statistics and helps visualization.</p>
<p>Sort the pixel level output, this is for visualize large images with limited memory usage.
<div class="highlight"><pre><span></span><code><span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.pixel.tsv.gz<span class="w"> </span><span class="c1"># j, X, Y, K1, ..., KJ, P1, ..., PJ, J=topk</span>
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.pixel.sorted.tsv.gz

<span class="nv">K</span><span class="o">=</span><span class="k">$(</span><span class="w"> </span><span class="nb">echo</span><span class="w"> </span><span class="nv">$model_id</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>sed<span class="w"> </span><span class="s1">&#39;s/nF\([0-9]\{1,\}\)\..*/\1/&#39;</span><span class="w"> </span><span class="k">)</span>
<span class="k">while</span><span class="w"> </span><span class="nv">IFS</span><span class="o">=</span><span class="s1">$&#39;\t&#39;</span><span class="w"> </span><span class="nb">read</span><span class="w"> </span>-r<span class="w"> </span>r_key<span class="w"> </span>r_val<span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">    </span><span class="nb">export</span><span class="w"> </span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">r_key</span><span class="si">}</span><span class="s2">&quot;</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">r_val</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="k">done</span><span class="w"> </span>&lt;<span class="w"> </span><span class="si">${</span><span class="nv">coor</span><span class="si">}</span>
<span class="nb">echo</span><span class="w"> </span>-e<span class="w"> </span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">xmin</span><span class="si">}</span><span class="s2">, </span><span class="si">${</span><span class="nv">xmax</span><span class="si">}</span><span class="s2">; </span><span class="si">${</span><span class="nv">ymin</span><span class="si">}</span><span class="s2">, </span><span class="si">${</span><span class="nv">ymax</span><span class="si">}</span><span class="s2">&quot;</span>

<span class="nv">offsetx</span><span class="o">=</span><span class="si">${</span><span class="nv">xmin</span><span class="si">}</span>
<span class="nv">offsety</span><span class="o">=</span><span class="si">${</span><span class="nv">ymin</span><span class="si">}</span>
<span class="nv">rangex</span><span class="o">=</span><span class="k">$(</span><span class="w"> </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;(</span><span class="si">${</span><span class="nv">xmax</span><span class="si">}</span><span class="s2"> - </span><span class="si">${</span><span class="nv">xmin</span><span class="si">}</span><span class="s2"> + 0.5)/1+1&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>bc<span class="w"> </span><span class="k">)</span>
<span class="nv">rangey</span><span class="o">=</span><span class="k">$(</span><span class="w"> </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;(</span><span class="si">${</span><span class="nv">ymax</span><span class="si">}</span><span class="s2"> - </span><span class="si">${</span><span class="nv">ymin</span><span class="si">}</span><span class="s2"> + 0.5)/1+1&quot;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>bc<span class="w"> </span><span class="k">)</span>
<span class="nv">bsize</span><span class="o">=</span><span class="m">2000</span>
<span class="nv">scale</span><span class="o">=</span><span class="m">100</span>
<span class="nv">header</span><span class="o">=</span><span class="s2">&quot;##K=</span><span class="si">${</span><span class="nv">K</span><span class="si">}</span><span class="s2">;TOPK=3\n##BLOCK_SIZE=</span><span class="si">${</span><span class="nv">bsize</span><span class="si">}</span><span class="s2">;BLOCK_AXIS=X;INDEX_AXIS=Y\n##OFFSET_X=</span><span class="si">${</span><span class="nv">offsetx</span><span class="si">}</span><span class="s2">;OFFSET_Y=</span><span class="si">${</span><span class="nv">offsety</span><span class="si">}</span><span class="s2">;SIZE_X=</span><span class="si">${</span><span class="nv">rangex</span><span class="si">}</span><span class="s2">;SIZE_Y=</span><span class="si">${</span><span class="nv">rangey</span><span class="si">}</span><span class="s2">;SCALE=</span><span class="si">${</span><span class="nv">scale</span><span class="si">}</span><span class="s2">\n#BLOCK\tX\tY\tK1\tK2\tK3\tP1\tP2\tP3&quot;</span>

<span class="o">(</span><span class="nb">echo</span><span class="w"> </span>-e<span class="w"> </span><span class="s2">&quot;</span><span class="si">${</span><span class="nv">header</span><span class="si">}</span><span class="s2">&quot;</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span>zcat<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-n<span class="w"> </span>+2<span class="w"> </span><span class="p">|</span><span class="w"> </span>perl<span class="w"> </span>-slane<span class="w"> </span><span class="s1">&#39;$F[0]=int(($F[1]-$offx)/$bsize) * $bsize; $F[1]=int(($F[1]-$offx)*$scale); $F[1]=($F[1]&gt;=0)?$F[1]:0; $F[2]=int(($F[2]-$offy)*$scale); $F[2]=($F[2]&gt;=0)?$F[2]:0; print join(&quot;\t&quot;, @F);&#39;</span><span class="w"> </span>--<span class="w"> </span>-bsize<span class="o">=</span><span class="si">${</span><span class="nv">bsize</span><span class="si">}</span><span class="w"> </span>-scale<span class="o">=</span><span class="si">${</span><span class="nv">scale</span><span class="si">}</span><span class="w"> </span>-offx<span class="o">=</span><span class="si">${</span><span class="nv">offsetx</span><span class="si">}</span><span class="w"> </span>-offy<span class="o">=</span><span class="si">${</span><span class="nv">offsety</span><span class="si">}</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-S<span class="w"> </span>4G<span class="w"> </span>-k1,1g<span class="w"> </span>-k3,3g<span class="w"> </span><span class="o">)</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>bgzip<span class="w"> </span>-c<span class="w"> </span>&gt;<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span>

tabix<span class="w"> </span>-f<span class="w"> </span>-s1<span class="w"> </span>-b3<span class="w"> </span>-e3<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span>
rm<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span>
</code></pre></div></p>
<p>Report differentially expressed genes. This is a naive pseudo-bulk chi-squared test, please view the results with caution.
<div class="highlight"><pre><span></span><code><span class="c1"># DE</span>
<span class="nv">max_pval_output</span><span class="o">=</span>1e-3
<span class="nv">min_fold_output</span><span class="o">=</span><span class="m">1</span>.5
<span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.posterior.count.tsv.gz
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.bulk_chisq.tsv
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/de_bulk.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--min_ct_per_feature<span class="w"> </span><span class="si">${</span><span class="nv">min_ct_per_feature</span><span class="si">}</span><span class="w"> </span>--max_pval_output<span class="w"> </span><span class="si">${</span><span class="nv">max_pval_output</span><span class="si">}</span><span class="w"> </span>--min_fold_output<span class="w"> </span><span class="si">${</span><span class="nv">min_fold_output</span><span class="si">}</span><span class="w"> </span>--thread<span class="w"> </span><span class="si">${</span><span class="nv">thread</span><span class="si">}</span>


<span class="c1"># Report (color table and top DE genes)</span>
<span class="nv">cmap</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/figure/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.rgb.tsv
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.factor.info.html
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/factor_report.py<span class="w"> </span>--path<span class="w"> </span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span><span class="w"> </span>--pref<span class="w"> </span><span class="si">${</span><span class="nv">prefix</span><span class="si">}</span><span class="w"> </span>--color_table<span class="w"> </span><span class="si">${</span><span class="nv">cmap</span><span class="si">}</span>
</code></pre></div></p>
<p>Generalize pixel level images representing the factorization result
<div class="highlight"><pre><span></span><code><span class="c1"># Make pixel level figures</span>
<span class="nv">cmap</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/figure/<span class="si">${</span><span class="nv">output_id</span><span class="si">}</span>.rgb.tsv
<span class="nv">input</span><span class="o">=</span><span class="si">${</span><span class="nv">output_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.pixel.sorted.tsv.gz
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.pixel.png
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/plot_pixel_full.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--color_table<span class="w"> </span><span class="si">${</span><span class="nv">cmap</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--plot_um_per_pixel<span class="w"> </span><span class="m">0</span>.5<span class="w"> </span>--full
</code></pre></div></p>
<p>Generate heatmaps for individual factors. If the data is very large, making all individual factor maps may take some time.</p>
<p>Generate everything in one run
<div class="highlight"><pre><span></span><code><span class="c1"># Make single factor heatmaps, plot_subbatch balances speed and memory ...</span>
<span class="c1"># batch size 8 should be safe for 7 or 14G in most cases</span>
<span class="nv">output</span><span class="o">=</span><span class="si">${</span><span class="nv">figure_path</span><span class="si">}</span>/sub/<span class="si">${</span><span class="nv">prefix</span><span class="si">}</span>.pixel
python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/plot_pixel_single.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--plot_um_per_pixel<span class="w"> </span><span class="m">0</span>.5<span class="w"> </span>--full<span class="w"> </span>--all
</code></pre></div></p>
<p>Alternatively, you can generate by batch
<div class="highlight"><pre><span></span><code><span class="nv">plot_subbatch</span><span class="o">=</span><span class="m">8</span>
<span class="nv">st</span><span class="o">=</span><span class="m">0</span>
<span class="nv">ed</span><span class="o">=</span><span class="k">$((</span><span class="nv">plot_subbatch</span><span class="o">+</span><span class="nv">st</span><span class="o">-</span><span class="m">1</span><span class="k">))</span>
<span class="k">while</span><span class="w"> </span><span class="o">[</span><span class="w"> </span><span class="si">${</span><span class="nv">st</span><span class="si">}</span><span class="w"> </span>-lt<span class="w"> </span><span class="si">${</span><span class="nv">K</span><span class="si">}</span><span class="w"> </span><span class="o">]</span><span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="o">[</span><span class="w"> </span><span class="si">${</span><span class="nv">ed</span><span class="si">}</span><span class="w"> </span>-gt<span class="w"> </span><span class="si">${</span><span class="nv">K</span><span class="si">}</span><span class="w"> </span><span class="o">]</span><span class="p">;</span><span class="w"> </span><span class="k">then</span>
<span class="w">        </span><span class="nv">ed</span><span class="o">=</span><span class="k">$((</span><span class="nv">K</span><span class="o">-</span><span class="m">1</span><span class="k">))</span>
<span class="w">    </span><span class="k">fi</span>
<span class="w">    </span><span class="nv">id_list</span><span class="o">=</span><span class="k">$(</span><span class="w"> </span>seq<span class="w"> </span><span class="si">${</span><span class="nv">st</span><span class="si">}</span><span class="w"> </span><span class="si">${</span><span class="nv">ed</span><span class="si">}</span><span class="w"> </span><span class="k">)</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="nv">$id_list</span>

<span class="w">    </span>python<span class="w"> </span><span class="si">${</span><span class="nv">gitpath</span><span class="si">}</span>/script/plot_pixel_single.py<span class="w"> </span>--input<span class="w"> </span><span class="si">${</span><span class="nv">input</span><span class="si">}</span><span class="w"> </span>--output<span class="w"> </span><span class="si">${</span><span class="nv">output</span><span class="si">}</span><span class="w"> </span>--id_list<span class="w"> </span><span class="si">${</span><span class="nv">id_list</span><span class="si">}</span><span class="w"> </span>--plot_um_per_pixel<span class="w"> </span><span class="m">0</span>.5<span class="w"> </span>--full
<span class="w">    </span><span class="nv">st</span><span class="o">=</span><span class="k">$((</span><span class="nv">ed</span><span class="o">+</span><span class="m">1</span><span class="k">))</span>
<span class="w">    </span><span class="nv">ed</span><span class="o">=</span><span class="k">$((</span><span class="nv">plot_subbatch</span><span class="o">+</span><span class="nv">st</span><span class="o">-</span><span class="m">1</span><span class="k">))</span>
<span class="k">done</span>
</code></pre></div></p>
<h2 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h2>
<p>In the above example the analysis outputs are stored in
<div class="highlight"><pre><span></span><code>${path}/analysis/${model_id} # examples/data/analysis/nF12.d_12
</code></pre></div></p>
<p>There is an html file reporting the color code and top genes of the inferred factors
<div class="highlight"><pre><span></span><code>nF12.d_12.decode.prj_12.r_4_5.factor.info.html
</code></pre></div></p>
<p>Pixel level visualizating
<div class="highlight"><pre><span></span><code>figure/nF12.d_12.decode.prj_12.r_4_5.pixel.png
</code></pre></div></p>
<p>Pixel level output is
<div class="highlight"><pre><span></span><code>nF12.d_12.decode.prj_12.r_4_5.pixel.sorted.tsv.gz
</code></pre></div></p>
<p>We store the top 3 factors and their corresponding posterior probabilities for each pixel in tab delimted text files.
As a temporary hack for accessing specific regions in large dataset faster, we divided the data along one axis (X or Y), sorted within each block by the other axis.
The first 3 lines of the file, starting with <code>##</code>, are metadata, the 4<sup>th</sup> line, starting with <code>#</code>, contains columns names.
To use the file as plain text, you can ignore this complication and read the file from the 4<sup>th</sup> line.</p>
<p>The first few lines of the file are as follows:</p>
<div class="highlight"><pre><span></span><code>##K=12;TOPK=3
##BLOCK_SIZE=2000;BLOCK_AXIS=X;INDEX_AXIS=Y
##OFFSET_X=6690;OFFSET_Y=6772;SIZE_X=676;SIZE_Y=676;SCALE=100
#BLOCK  X       Y       K1      K2      K3      P1      P2      P3
0       10400   360     2       1       8       9.07e-01        9.27e-02        2.61e-13
0       10669   360     2       1       8       9.36e-01        6.37e-02        4.20e-08
0       10730   360     2       1       8       8.85e-01        1.15e-01        1.83e-05
</code></pre></div>
<p>The 4<sup>th</sup> line contains the column names. From the 5<sup>th</sup> line on, each line contains the information for one pixel with coordinates <code>(X, Y)</code>, the top 3 factors indicated by <code>K1, K2, K3</code> and their corresponding posterior probabilities <code>P1, P2, P3</code>. Factors are 0-indexed.</p>
<p>The 1<sup>st</sup> line indicates that the data is from a model with 12 factors (<code>K=12</code>) and we store the top 3 factors for each pixel (<code>TOPK=3</code>).</p>
<p>The 2<sup>nd</sup> line indicates that the data is separated into blocks by the X axis (<code>BLOCK_AXIS=X</code>) with block size 2000<span class="arithmatex">\(\mu m\)</span> (<code>BLOCK_SIZE=2000</code>), then within each block the data is sorted by the Y axis (<code>INDEX_AXIS=Y</code>).
The block IDs (first column in the file) are integer multiples of the block size (in <span class="arithmatex">\(\mu m\)</span>), i.e. the 1<sup>st</sup> block, with <span class="arithmatex">\(X \in [0, 2000)\)</span> have block ID 0, the 2<sup>nd</sup> block, with <span class="arithmatex">\(X \in [2000, 4000)\)</span> have block ID 2000, etc.</p>
<p>The 3<sup>rd</sup> line describes the translation between the stored cooredinates and the physical coordinates in <span class="arithmatex">\(\mu m\)</span>.
Take <code>(X, Y)</code> as a pixel coordinates read from the file, the physical coordinates in <span class="arithmatex">\(\mu m\)</span> is <code>(X / SCALE + OFFSET_X, Y / SCALE + OFFSET_Y)</code>.
In this above example, the raw data from Vizgen MERSCOPE mouse liver data contains negative coordinates, but for convineince we shifted all coordinates to positive. <code>SIZE_X</code> and <code>SIZE_Y</code> record the size of the raw data in <span class="arithmatex">\(\mu m\)</span>.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.indexes", "navigation.sections"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.1e8ae164.min.js"></script>
      
    
  </body>
</html>